### PR TITLE
Basic flac support

### PR DESCRIPTION
**Pull request by rillian**
Basic support for the proposed FLAC encapsulation.

### CODE REVIEW

File: mp4parse/src/lib.rs [hunk 1]
```diff
@@ -1458,6 +1528,13 @@ fn be_u16<T: ReadBytesExt>(src: &mut T) -> Result<u16> {
     src.read_u16::<byteorder::BigEndian>().map_err(From::from)
 }
 
+fn be_u24<T: ReadBytesExt>(src: &mut T) -> Result<u32> {
+    let a = try!(src.read_u8()) as u32;
+    let b = try!(src.read_u8()) as u32;
+    let c = try!(src.read_u8()) as u32;
+    Ok( a << 16 | b << 8 | c)
```
Comment (kinetiknz) [inline mp4parse/src/lib.rs]: Space between Ok( and a

File: mp4parse/src/tests.rs [hunk 2]
```diff
@@ -442,6 +442,104 @@ fn read_hdlr_zero_length_name() {
...
+         .append_bytes(&make_dfla(FlacBlockType::StreamInfo, true,
+                                  &flac_streaminfo(), FlacBlockLength::Correct)
+         .into_inner())
+    });
+    let mut iter = super::BoxIter::new(&mut stream);
+    let mut stream = iter.next_box().unwrap().unwrap();
+    let mut track = super::Track::new(0);
+    let r = super::read_audio_desc(&mut stream, &mut track);
+    r.unwrap();
+    //assert!(r.is_ok());
```
Comment (kinetiknz) [inline mp4parse/src/tests.rs]: Remove commented code?

File: mp4parse_capi/src/lib.rs [hunk 3]
```diff
@@ -436,6 +439,9 @@ pub unsafe extern fn mp4parse_get_track_audio_info(parser: *mut mp4parse_parser,
             (*info).codec_specific_config.length = v.len() as u32;
             (*info).codec_specific_config.data = v.as_ptr();
         }
+        AudioCodecSpecific::FLACSpecificBox(ref _flac) => {
```
Comment (kinetiknz) [inline mp4parse_capi/src/lib.rs]: FLACSpecificBox(_) might be clearer

File: mp4parse/tests/public.rs [hunk 4]
```diff
@@ -74,6 +74,10 @@ fn public_api() {
                         assert!(v.len() > 0);
                         "ES"
                     }
+                    mp4::AudioCodecSpecific::FLACSpecificBox(_flac) => {
```
Comment (kinetiknz) [inline mp4parse/tests/public.rs]: FLACSpecificBox(_) might be clearer

File: mp4parse/src/lib.rs [hunk 5]
```diff
@@ -1075,11 +1090,57 @@ fn read_vpcc<T: Read>(src: &mut BMFFBox<T>) -> Result<VPxConfigBox> {
...
+    let mut blocks = Vec::new();
+    while src.bytes_left() > 0 {
+        let block = try!(read_flac_metadata(src));
+        blocks.push(block);
+    }
+    // The box must have at least one meta block, and the first block
+    // must be the METADATA_BLOCK_STREAMINFO
+    if blocks.len() < 1 {
+        return Err(Error::InvalidData("FLACSpecificBox missing metadata"));
+    }
```
Comment (kinetiknz) [inline mp4parse/src/lib.rs]: } and else if on the same line

File: mp4parse/src/lib.rs [hunk 6]
```diff
@@ -1075,11 +1090,57 @@ fn read_vpcc<T: Read>(src: &mut BMFFBox<T>) -> Result<VPxConfigBox> {
...
+        return Err(Error::Unsupported("unknown dfLa (FLAC) version"));
+    }
+    let mut blocks = Vec::new();
+    while src.bytes_left() > 0 {
+        let block = try!(read_flac_metadata(src));
+        blocks.push(block);
+    }
+    // The box must have at least one meta block, and the first block
+    // must be the METADATA_BLOCK_STREAMINFO
+    if blocks.len() < 1 {
```
Comment (kinetiknz) [inline mp4parse/src/lib.rs]: blocks.is_empty()

Comment (rillian) [inline mp4parse/src/lib.rs]: Oh, much better, thanks.

File: mp4parse/src/lib.rs (hunk 1 above)
Comment (rillian) [inline mp4parse/src/lib.rs]: Thanks

File: mp4parse/src/tests.rs (hunk 2 above)
Comment (rillian) [inline mp4parse/src/tests.rs]: Yes. The assert!() was equivalent here, but .unwrap() gives a better error message.

File: mp4parse_capi/src/lib.rs (hunk 3 above)
Comment (rillian) [inline mp4parse_capi/src/lib.rs]: yes

**Pull request merged**